module2/module2.js

```diff
@@ -0,0 +1,196 @@
+      return function(c) {
+        return function(d) {
+          return f(a, b, c, d);
+        };
+      };
+    };
+  };
+};
+
+function joinFunction(...args) {
+  return args.join(' ');
+}
+
+const mergeWords =  curry(joinFunction);
+
+describe('mergeWords', () => {
+  it('should merge words into sentence', () => {
+    const result = mergeWords('GNU')('is')('not')('Unix.'); //removed function call, is it ok?
+
+    expect(result).to.equal('GNU is not Unix.');
+  });  
+});
+
+// ==== Every/Some ====
+const goodUsers = [
+    { id: 1 },
+    { id: 2 },
+    { id: 3 },
+];
+
+// `checkUsersValid` is the function you'll define
+const checkUsersValid = (goodUsers) => {
+  return function(usersToTest) {
+    let absentUsers = usersToTest.filter( (user) => filterAbsentUser(user, goodUsers))
+    return absentUsers.length === 0; 
+  };
+};
+
+function filterAbsentUser(user, goodUsers) {
+  if (goodUsers.some(e => e.id === user.id)) {
+    return false;
+  }
+  return true;
+}
+
+const testAllValid = checkUsersValid(goodUsers);
+
+
+describe('checkUsersValid', () => {
+  it('#case 1', () => {
+    const result = testAllValid([
+      { id: 2 },
+      { id: 1 },
+    ]);
+
+    expect(result).to.equal(true);
+  });
+  it('#case 2', () => {
+    const result = testAllValid([
+      { id: 2 },
+      { id: 4 },
+      { id: 1 },
+    ]);
+
+    expect(result).to.equal(false);
+  });
+});
+
+// ==== Reduce ====
+const reducer = (accumulator, currentValue) => { 
+  if(accumulator[currentValue] === undefined) {
+    accumulator[currentValue] = 1;
+    return accumulator;
+  }
+  accumulator[currentValue] =accumulator[currentValue]+ 1;
+  return accumulator;
+}
+
+const countWords = (words) => {
+  return words.reduce(reducer, {})
+};
+
+describe('reduce', () => {
+  it('should return an object which contains the number of times each string occured in the array', () => {
+    const inputWords = ['Apple', 'Banana', 'Apple', 'Durian', 'Durian', 'Durian'];
+    const result = countWords(inputWords);
+    expect(result).to.eql({ 'Apple': 2, 'Banana': 1, 'Durian': 3 });
+  });  
+});
+
+// ==== Palindrome ====
+function reverse(word) {
+  return word.split('').reverse().join('');
+}
+
+const isPalindrome = (word) => {
+  return word === reverse(word) ? 'The entry is a palindrome' : 'Entry is not a palindrome';
+}
+
+describe('isPalindrome', () => {
+  it('should check whether a passed string is palindrome or not', () => {
+    expect(isPalindrome('madam')).to.equal('The entry is a palindrome');
+    expect(isPalindrome('fox')).to.equal('Entry is not a palindrome');
+  });
+});
+
+// ==== Recursion ====
+const factorial = (n) => {
+  if(n < 0) return null;
+  if(n===0 || n===1) {
+    return 1;
+  }
+  return n*factorial(n - 1);
+};
+
+const amountToCoins = (n, coins, result=[]) => {    
+  const fittedCoins = coins.filter( c => n/c >= 1);
+  const maxCoin = Math.max(...fittedCoins);
+  if (maxCoin === undefined || n === 0) return result;
```

Comment:
You don't have to check on undefined. It should be done just by using the value you check.
if(!maxCoin) does the same.